components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodCLIService.kt

```diff
@@ -64,10 +65,30 @@ class GitpodCLIService : RestService() {
             }
         }
         if (operation == "preview") {
-            val url = getStringParameter("url", urlDecoder)
+            var url = getStringParameter("url", urlDecoder)
+
             if (url.isNullOrBlank()) {
                 return "url is missing"
             }
+
+            // When it's auto-forwarding ports, we need to consider Gitpod Tasks running "gp preview".
+            // For example: gp preview --external $(gp url 3000)
+            if (manager.isAutoForwardingPorts) {
```

Comment:
if a user wants to access exposed API then we should not translate it to local.

We should mimic logic of VS Code [[1](https://github.com/gitpod-io/openvscode-server/blob/9d68ec908452a065efdd6e8430c67b2898e4f038/src/vs/gitpod/browser/workbench/workbench.ts#L908-L926)] and apply it in similar situations.

Reply:
Let me check.

Reply:
@akosyakov, I didn't understand the usage of tunnels and how I should use them here.

The port forwarding of JetBrains is not related to the tunnels used by the Local Companion App, right?

Could you clarify? Or we can sync tomorrow.

Reply:
In this code, tunnel is basically like port forwarding in JB, so the logic is:
- if a user is trying to open URI which is local [1](https://github.com/gitpod-io/openvscode-server/blob/9d68ec908452a065efdd6e8430c67b2898e4f038/src/vs/platform/tunnel/common/tunnel.ts#L134)] then apply port forwarding mapping
- otherwise just open URI 

So only localhost URL can be translated, not localhost URIs are never translated.

Try to play with VS Code Browser + Local Companion for following use cases:
- `gp preview localhost:3000` when let's say 3000 is forwarded on 3001
- and then `gp preview $(gp url 3000)` 

it should work the same way for JB and VS Code

Reply:
ℹ️ Haven't had the time to update it yet, but I will.

Reply:
I've confirmed that gp-cli's `preview` command already converts the 'localhost' to the workspace URL:
https://github.com/gitpod-io/gitpod/blob/bffbd3216205d568e9820a7eae30d2af8ea855ba/components/gitpod-cli/cmd/preview.go#L27-L32

So even now (without the changes from this PR), notifications display the remote address (e.g. `https://8080-gitpodio-gitpod-zufs0q98iqf.ws-eu54.gitpod.io`) when using `gp preview http://localhost:8080` or `gp preview $(gp url 8080)`:
<img width="416" alt="image" src="https://user-images.githubusercontent.com/418083/182356572-d7f45e7c-00e0-4a19-9065-58c09dbcb99c.png">

As it's never receiving a localhost URL, I think it doesn't worth changing [GitpodCLIService.kt](https://github.com/gitpod-io/gitpod/pull/11081/files#diff-81d1cb74ab7142c7fba14c9bb140bb6a38eedf7d9e72a5f10065b130f95d2801) in this PR.

Reply:
let's create an issue for it, we should move translation to IDEs from gp cli

Reply:
Issue created:
- https://github.com/gitpod-io/gitpod/issues/11849